src/pybind/mgr/dashboard/controllers/docs.py

```diff
@@ -306,21 +308,21 @@ def _gen_paths(cls, all_endpoints, baseUrl):
 
         return paths
 
-    def _gen_spec(self, all_endpoints=False, baseUrl=""):
+    def _gen_spec(self, all_endpoints=False, base_url=""):
         if all_endpoints:
-            baseUrl = ""
+            base_url = ""
 
         host = cherrypy.request.base
         host = host[host.index(':')+3:]
         logger.debug("DOCS: Host: %s", host)
 
-        paths = self._gen_paths(all_endpoints, baseUrl)
+        paths = self._gen_paths(all_endpoints, base_url)
 
-        if not baseUrl:
-            baseUrl = "/"
+        if not base_url:
+            base_url = "/"
 
         scheme = 'https'
-        ssl = mgr.get_localized_module_option('ssl', 'True')
+        ssl = str_to_bool(mgr.get_localized_module_option('ssl', 'True'))
```

Comment:
@votdev did the return type of mgr.get_localized_module_option changed again? When we merged Tina PR, this function was returning a `bool` typed value, so there was no need to convert it.

Reply:
You're right, this is not necessary anymore. Sorry, will remove that from the PR.

Reply:
Be careful @votdev. I removed all `str_to_bool`'s in a previous PR and had to put them all back afterwards. The reason? Some config options, as port, address and ssl are _usually_ set from command line before the ceph-mgr is loaded and any Dashboard option schema is in place (that's why they require `--force` in `ceph config set`). But the config store has no hints on whether that `ssl` setting is `CephBool`, `CephInt` or `CephString` (which is the default), so eventually `ssl` ends up stored as the string `"True"`.

The setback of all of this is that both approaches are incompatible, so if the `ssl` option is set after the Ceph-mgr has loaded the Ceph-dashboard it'll be treated as a Python `bool`, and hence the `str_to_bool` will fail:
```py
>>> strtobool(True)
Traceback (most recent call last):
...
AttributeError: 'bool' object has no attribute 'lower'
```
I exchanged an e-mail with @liewegas about this matter on whether it would make sense to allow  `ceph config set` to enforce a given type, but we didn't dig further. Maybe we could move this type-handling logic to the Python MgrModule or have a 'best-effort' type casting in `mon/ConfigMonitor.cc` when `forced` and `Option::parse_value()` for `TYPE_{BOOL,INT,UINT}`.

Reply:
I'm wondering because ``get_localized_module_option`` end up in the Ceph Python extension which converts finally the result to the configured MODULE_OPTIONS type, see 

- https://github.com/ceph/ceph/blob/master/src/mgr/BaseMgrModule.cc#L399
- https://github.com/ceph/ceph/blob/master/src/mgr/ActivePyModules.cc#L528

Reply:
> Some config options, as port, address and ssl are _usually_ set from command line before the ceph-mgr is loaded and any Dashboard option schema is in place

Argghh

Reply:
@epuertat i think i may have missed or misunderstood that earlier discussion.  the config-key store in the mon always stores strings.. it's the mgr/Py* code that maps that to the right python type when it is being fetched, and at that stage we *always* know the proper type.  So i'm not sure how we could end up getting something other than a python bool...

Reply:
Is the problem that the value might be a string (e.g., 'foo') which can't be cast to a CephInt, and so the code returns a string?  That's easy to fix (we should return None if the string can't be cast to the proper type!).

Reply:
@liewegas let me re-test it. Right after #25597 was merged I tried removing explicit type casting (`int()`) from dashboard `server_port`, added the `type: 'int'` to the MODULE_OPTIONS and the method `get_module_option()` was still returning a string. I found that the behaviour was different depending on whether the option was forced-set before Mgr started or set afterwards.

Reply:
Please let us know about your test result. In the meanwhile i will still keep the ``str_to_bool`` call. At the end this does not hurt if we keep this, even if it is not necessary at the end.

Reply:
Just tried:
```diff
diff --git a/src/pybind/mgr/dashboard/module.py b/src/pybind/mgr/dashboard/module.py
index 9a6f374832..10792e2785 100644
--- a/src/pybind/mgr/dashboard/module.py
+++ b/src/pybind/mgr/dashboard/module.py
@@ -150,7 +150,7 @@ class CherryPyConfig(object):
         config = {
             'engine.autoreload.on': False,
             'server.socket_host': server_addr,
-            'server.socket_port': int(server_port),
+            'server.socket_port': server_port,
             'error_page.default': json_error_page,
             'tools.request_logging.on': True,
             'tools.gzip.on': True,
```
with a localized port (`mgr/dashboard/x/server_port=11000`), and the error persists:
```bash
2019-03-08 09:06:02.133 7f3b952bf700 -1 Traceback (most recent call last):
  File "/ceph/src/pybind/mgr/dashboard/module.py", line 340, in serve
    cherrypy.engine.start()
  File "/usr/lib/python3.7/site-packages/cherrypy/process/wspbus.py", line 269, in start
    raise e_info
  File "/usr/lib/python3.7/site-packages/cherrypy/process/wspbus.py", line 254, in start
    self.publish('start')
  File "/usr/lib/python3.7/site-packages/cherrypy/process/wspbus.py", line 234, in publish
    raise exc
cherrypy.process.wspbus.ChannelFailures: TypeError('an integer is required (got type str)')
```

But if I use a non-localized setting:
```diff
diff --git a/src/pybind/mgr/dashboard/module.py b/src/pybind/mgr/dashboard/module.py
index 9a6f374832..7324dc271d 100644
--- a/src/pybind/mgr/dashboard/module.py
+++ b/src/pybind/mgr/dashboard/module.py
@@ -127,7 +127,7 @@ class CherryPyConfig(object):
         if not ssl:
             def_server_port = 8080
 
-        server_port = self.get_localized_module_option('server_port', def_server_port)
+        server_port = self.get_module_option('server_port', def_server_port)
         if server_addr is None:
             raise ServerConfigException(
                 'no server_addr configured; '
@@ -150,7 +150,7 @@ class CherryPyConfig(object):
         config = {
             'engine.autoreload.on': False,
             'server.socket_host': server_addr,
-            'server.socket_port': int(server_port),
+            'server.socket_port': server_port,
             'error_page.default': json_error_page,
             'tools.request_logging.on': True,
             'tools.gzip.on': True,
diff --git a/src/vstart.sh b/src/vstart.sh
index 02977b5e36..9c1bbe960d 100755
--- a/src/vstart.sh
+++ b/src/vstart.sh
@@ -789,7 +789,7 @@ start_mgr() {
 EOF
 
             if $with_mgr_dashboard ; then
-                ceph_adm config set mgr mgr/dashboard/$name/server_port $MGR_PORT --force
+                ceph_adm config set mgr mgr/dashboard/server_port $MGR_PORT --force
                 if [ $mgr -eq 1 ]; then
                     DASH_URLS="https://$IP:$MGR_PORT"
                 else
```
everything works fine. So this means that the issue comes from the localized handling.

Reply:
So, let's get rid of `str_to_bool` (and don't forget to change default `'True'` to `True`).

Reply:
> So, let's get rid of `str_to_bool` (and don't forget to change default `'True'` to `True`).

That's not possible, the default value must be a string because the Cython implementation of ``ceph_get_module_option`` is requesting this: https://github.com/ceph/ceph/blob/master/src/mgr/BaseMgrModule.cc#L403

Reply:
> Just tried:
> 
> ```diff
> diff --git a/src/pybind/mgr/dashboard/module.py b/src/pybind/mgr/dashboard/module.py
> index 9a6f374832..10792e2785 100644
> --- a/src/pybind/mgr/dashboard/module.py
> +++ b/src/pybind/mgr/dashboard/module.py
> @@ -150,7 +150,7 @@ class CherryPyConfig(object):
>          config = {
>              'engine.autoreload.on': False,
>              'server.socket_host': server_addr,
> -            'server.socket_port': int(server_port),
> +            'server.socket_port': server_port,
>              'error_page.default': json_error_page,
>              'tools.request_logging.on': True,
>              'tools.gzip.on': True,
> ```
> 
> with a localized port (`mgr/dashboard/x/server_port=11000`), and the error persists:
> 
> ```shell
> 2019-03-08 09:06:02.133 7f3b952bf700 -1 Traceback (most recent call last):
>   File "/ceph/src/pybind/mgr/dashboard/module.py", line 340, in serve
>     cherrypy.engine.start()
>   File "/usr/lib/python3.7/site-packages/cherrypy/process/wspbus.py", line 269, in start
>     raise e_info
>   File "/usr/lib/python3.7/site-packages/cherrypy/process/wspbus.py", line 254, in start
>     self.publish('start')
>   File "/usr/lib/python3.7/site-packages/cherrypy/process/wspbus.py", line 234, in publish
>     raise exc
> cherrypy.process.wspbus.ChannelFailures: TypeError('an integer is required (got type str)')
> ```
> 
> But if I use a non-localized setting:
> 
> ```diff
> diff --git a/src/pybind/mgr/dashboard/module.py b/src/pybind/mgr/dashboard/module.py
> index 9a6f374832..7324dc271d 100644
> --- a/src/pybind/mgr/dashboard/module.py
> +++ b/src/pybind/mgr/dashboard/module.py
> @@ -127,7 +127,7 @@ class CherryPyConfig(object):
>          if not ssl:
>              def_server_port = 8080
>  
> -        server_port = self.get_localized_module_option('server_port', def_server_port)
> +        server_port = self.get_module_option('server_port', def_server_port)
>          if server_addr is None:
>              raise ServerConfigException(
>                  'no server_addr configured; '
> @@ -150,7 +150,7 @@ class CherryPyConfig(object):
>          config = {
>              'engine.autoreload.on': False,
>              'server.socket_host': server_addr,
> -            'server.socket_port': int(server_port),
> +            'server.socket_port': server_port,
>              'error_page.default': json_error_page,
>              'tools.request_logging.on': True,
>              'tools.gzip.on': True,
> diff --git a/src/vstart.sh b/src/vstart.sh
> index 02977b5e36..9c1bbe960d 100755
> --- a/src/vstart.sh
> +++ b/src/vstart.sh
> @@ -789,7 +789,7 @@ start_mgr() {
>  EOF
>  
>              if $with_mgr_dashboard ; then
> -                ceph_adm config set mgr mgr/dashboard/$name/server_port $MGR_PORT --force
> +                ceph_adm config set mgr mgr/dashboard/server_port $MGR_PORT --force
>                  if [ $mgr -eq 1 ]; then
>                      DASH_URLS="https://$IP:$MGR_PORT"
>                  else
> ```
> 
> everything works fine. So this means that the issue comes from the localized handling.

This behavior is really strange and i don't know how this could happen because in the end ``get_module_option`` and ``get_localized_module_option`` end up in calling ``ceph_get_module_option`` in the Mgr Python extension.

Reply:
The localized method has a problem after the MODULE_OPTIONS feature has been introduced. Maybe that's the reason for this. PR https://github.com/ceph/ceph/pull/26736 is fixing the localized problem.

Reply:
> > So, let's get rid of `str_to_bool` (and don't forget to change default `'True'` to `True`).
> 
> That's not possible, the default value must be a string because the Cython implementation of `ceph_get_module_option` is requesting this: https://github.com/ceph/ceph/blob/master/src/mgr/BaseMgrModule.cc#L403

Sorry, that's incorrect. You can pass everything you want. The function parses the specified default value as string for further use.

Reply:
@epuertat I investigated your issue and you're right, the MODULE_OPTIONS auto-convertion does not work work ``get_localized_module_options``. It seems the Python extension is not clever enough to do so. I assume a key like ``x/server_port`` is not handled properly like ``server_port``. Maybe this is something we should raise in a tracker issue.

Reply:
Maybe we simply have to strip the key (``x/foo`` => ``foo``) in ``PyModule::get_typed_option_value``, see https://github.com/ceph/ceph/blob/master/src/mgr/PyModule.cc#L627.

Reply:
I've opened a tracker issue for that: https://tracker.ceph.com/issues/38639

Reply:
thanks, @votdev! I'll join looking at that issue as soon as I fix the lazy loading weirdness I'm struggling with :sweat_smile: